### PR TITLE
Exporting Solidity AST in a JSON format

### PR DESCRIPTION
**Pull request by LefterisJP**
This is a first version.
Actual content of the JSON may change as usage by outside tools becomes more clear.

### CODE REVIEW

File: libsolidity/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+	m_astJson["children"] = children;
+	m_jsonNodePtrs.push(&m_astJson["children"]);
+}
+
+void ASTJsonConverter::print(ostream& _stream)
+{
+	m_ast->accept(*this);
+	_stream << m_astJson;
+}
+
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: blank line

File: libsolidity/ASTJsonConverter.cpp [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+	addJsonNode("Literal",
+				{make_pair("string", (tokenString) ? tokenString : "null"),
+					make_pair("value", _node.getValue()),
+					make_pair("type", getType(_node))});
+	return true;
+}
+
+void ASTJsonConverter::endVisit(ImportDirective const&)
+{
+
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: blank line

File: libsolidity/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+	goUp();
+}
+
+void ASTJsonConverter::endVisit(VariableDeclaration const&)
+{
+}
+
+void ASTJsonConverter::endVisit(TypeName const&)
+{
+
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: blank lines...

File: libsolidity/ASTJsonConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+void ASTJsonConverter::endVisit(Literal const&)
+{
+
+}
+
+string const ASTJsonConverter::getType(Expression const& _expression)
+{
+	return  (_expression.getType()) ? _expression.getType()->toString() : "Unknown";
+}
+
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: blank line

File: libsolidity/ASTJsonConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+								   initializer_list<pair<string const, string const>> _list,
+								   bool _hasChildren = false)
+{
+	Json::Value node;
+
+	node["name"] = _nodeName;
+	if (_list.size() !=0)
+	{
+		Json::Value attrs;
+		for (auto &e: _list)
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: `auto& e` space in wrong place.

File: libsolidity/ASTJsonConverter.cpp [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+
+bool ASTJsonConverter::visit(ImportDirective const& _node)
+{
+	addJsonNode("Import", { make_pair("file", _node.getIdentifier())});
+	return true;
+}
+
+bool ASTJsonConverter::visit(ContractDefinition const& _node)
+{
+	addJsonNode("Contract", { make_pair("name", _node.getName())}, true);
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: `) },` - note symmetrical spacing.

File: libsolidity/ASTJsonConverter.cpp [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+	addJsonNode("ParameterList", {}, true);
+	return true;
+}
+
+bool ASTJsonConverter::visit(FunctionDefinition const& _node)
+{
+	addJsonNode("Function",
+				{ make_pair("name", _node.getName()),
+					make_pair("public", boost::lexical_cast<std::string>(_node.isPublic())),
+					make_pair("const", boost::lexical_cast<std::string>(_node.isDeclaredConst()))},
```
Comment (gavofyork) [inline libsolidity/ASTJsonConverter.cpp]: same (probably below, too). don't cramp the `}`.

File: solc/CommandLineInterface.cpp [hunk 8]
```diff
@@ -339,20 +344,44 @@ bool CommandLineInterface::processInput()
 	return true;
 }
 
-void CommandLineInterface::actOnInput()
+void CommandLineInterface::handleAst(std::string const& _argStr)
 {
+	std::string title;
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: remove "std::"

File: libsolidity/ASTJsonConverter.h [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+{
+
+/**
+ * Converter of the AST into JSON format
+ */
+class ASTJsonConverter: public ASTConstVisitor
+{
+public:
+	/// Create a converter for the given abstract syntax tree. If the source is specified,
+	/// the corresponding parts of the source are printed with each node.
```
Comment (chriseth) [inline libsolidity/ASTJsonConverter.h]: Does it really print the source?

File: libsolidity/ASTJsonConverter.h [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	void endVisit(Identifier const&) override;
+	void endVisit(ElementaryTypeNameExpression const&) override;
+	void endVisit(Literal const&) override;
+
+private:
+	void addKeyValue(Json::Value& _obj, std::string const& _key, std::string const& _val);
+	void addJsonNode(std::string const& _nodeName,
+					 std::initializer_list<std::pair<std::string const, std::string const>> _list,
+					 bool _hasChildren);
+	std::string const getType(Expression const& _expression);
```
Comment (chriseth) [inline libsolidity/ASTJsonConverter.h]: Do you really want to return a const string?

File: libsolidity/ASTJsonConverter.h [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	inline void goUp()
+	{
+		if (m_jsonNodePtrs.empty())
+			BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Internal error"));
```
Comment (chriseth) [inline libsolidity/ASTJsonConverter.h]: I think this errinfo_comment can also be omitted.

File: libsolidity/ASTJsonConverter.cpp [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+}
+
+void ASTJsonConverter::addJsonNode(string const& _nodeName,
+								   initializer_list<pair<string const, string const>> _list,
+								   bool _hasChildren = false)
+{
+	Json::Value node;
+
+	node["name"] = _nodeName;
+	if (_list.size() !=0)
```
Comment (chriseth) [inline libsolidity/ASTJsonConverter.cpp]: space

File: libsolidity/ASTJsonConverter.cpp [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+	{
+		Json::Value attrs;
+		for (auto &e: _list)
+			addKeyValue(attrs, e.first, e.second);
+		node["attributes"] = attrs;
+	}
+
+	(*m_jsonNodePtrs.top()).append(node);
```
Comment (chriseth) [inline libsolidity/ASTJsonConverter.cpp]: Why not `m_jsonNodePtr.top()->append(node)`?

File: libsolidity/ASTJsonConverter.cpp [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+
+	if (_hasChildren)
+	{
+		Json::Value& addedNode = (*m_jsonNodePtrs.top())[m_jsonNodePtrs.top()->size() - 1];
```
Comment (chriseth) [inline libsolidity/ASTJsonConverter.cpp]: `m_jsonNodePtrs.top()->back()`?

File: libsolidity/ASTJsonConverter.h (hunk 9 above)
Comment (LefterisJP) [inline libsolidity/ASTJsonConverter.h]: *couch* copy paste *cough*

File: libsolidity/ASTJsonConverter.h (hunk 10 above)
Comment (LefterisJP) [inline libsolidity/ASTJsonConverter.h]: why not?

File: libsolidity/ASTJsonConverter.h (hunk 11 above)
Comment (LefterisJP) [inline libsolidity/ASTJsonConverter.h]: Yeah this is basicaly a developing time assert. Should never happen if the code is correct.

File: libsolidity/ASTJsonConverter.cpp (hunk 13 above)
Comment (LefterisJP) [inline libsolidity/ASTJsonConverter.cpp]: This is the main reason why I spend a hell of a lot of time debugging this code. There is an internal jsoncpp assertion that checks for the type of the Json::Value and triggers an internal library assert if it's a pointer. I have no idea why.

File: libsolidity/ASTJsonConverter.cpp (hunk 14 above)
Comment (LefterisJP) [inline libsolidity/ASTJsonConverter.cpp]: `Json::Value` has no function called `back()` unfortunately.

**Pull request merged**